### DISCUSSION TITLE
Possible way to reduce TTI

### DISCUSSION THREAD

**Discussion by janbiedermann** [opal/opal#2214] [General]
To reduce startup time, TTI, not asset size, it could be possible to reduce the code executed at startup. Currently, when we `require` something it means 2 things:
1. include the file the assets
2. execute file during startup/load

What i mean is decoupling 1. and 2. to something like that:
1. include the file in the assets
2. load when constant is actually requested. (following autoloader convention)

This way Opal would not have to execute all of stdlib before doing something useful, like showing something to click on in the browser.

Of course, this would break for things that reopen constants, so for these the current `require` behavior would be required. But for anything else i could imagine this could bring a boost.

Maybe this could be implemented in a backwards compatible way by passing an option to require:
`require 'promise', :auto`

And of course then tagging all requires in stdlib etc.

❓

**Comment by janbiedermann** [comment]
Just came to my mind, when opening a class, maybe it would be good then to first check if the original class can be autloaded, maybe that way this can be solved.

**Comment by elia** [comment]
I think that's a great idea, and there had been discussion around this concept in the past.

> This way Opal would not have to execute all of stdlib before doing something useful, like showing something to click on in the browser.

To be fair Opal doesn't require the stdlib right away, and, unlike JS modules, you can place your requires everywhere you want. For example you can require `Promise` just before using it and delay to code loading.

But the concept is still very useful and doing that for parts of the corelib (which is loaded in full when you `require 'opal'`) would be relatively easy. We could simply add `opal/autoload` which instead of requiring everything sets up autoloads for the more optional libraries.

I did this quick experiment, but I've found quite difficult to gauge differences, even slowing down the CPU 6x because stuff like loading `string/encoding` the difference between runs was wide (like ~40ms vs ~70ms), so this needs some more serious research.

Here's the code I used just FYI

```rb
# opal/opal/autoload.rb

require 'opal/base'
require 'opal/mini'

require 'corelib/kernel/format'
require 'corelib/string/encoding'
autoload :Math, 'corelib/math'
autoload :Complex, 'corelib/complex'
autoload :Rational, 'corelib/rational'
autoload :Time, 'corelib/time'
autoload :Struct, 'corelib/struct'
require 'corelib/io'
require 'corelib/main'
autoload :Dir, 'corelib/dir'
autoload :File, 'corelib/file'
autoload :Process, 'corelib/process'
require 'corelib/random'
require 'corelib/random/mersenne_twister.js'

require 'corelib/unsupported'
```

and then tested on chrome with these 2 commands:

```
bin/opal -O -e'require "opal/autoload"' -Rserver
bin/opal -O -e'require "opal"' -Rserver --runner-options='{"port":3001}'

> Just came to my mind, when opening a class, maybe it would be good then to first check if the original class can be autoloaded, maybe that way this can be solved.

I would say that in this case the burden of explicitly requiring things would be on the file reopening the class. That's both very simple and doesn't deviate from plain Ruby.

The other potential issue I see is that some of the corelib/stdlib files doesn't just declare the constant but also add methods on Kernel and elsewhere (e.g. `corelib/complex.rb`), that may be an acceptable tradeoff, or we'll need to move the core extensions to different files and still allow autoloading for the main constant.

**Comment by janbiedermann** [comment]
Sorry, i meant corelib.
There is Lighthouse to measure https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk

**Comment by janbiedermann** [comment]
Besides `autoload`, what could help larger frameworks/project that use `require_tree` (isomorfeus, hyperstack) would then be a `auto_require_tree` that loads code on demand. But isomorfeus and hyperstacks constant are offset by the tree root, for example:
```
require_tree 'components'
```
with a file 'components/my_component.rb' creates a top level constant:
```
MyComponent
```
(opal-autoloader is used for resolving constants here and allows for specifying the offset directory)

**Comment by janbiedermann** [comment]
I implemented a `require_tree 'component', :autoload`
see https://github.com/opal/opal/pull/1976#issuecomment-542845741

Effect:
![Bildschirmfoto 2019-10-16 um 20 43 22](https://user-images.githubusercontent.com/30776880/66950350-7c04fa00-f058-11e9-80b2-aa5a25e07133.png)

That is for a small app, but there are large apps where require_tree takes seconds and now will only take a few ms.

Lets see how that patch works out.
(Patch would need some polish for later on).

**Comment by janbiedermann** [comment]
It further shows that a browser, parsing "literal" javascript takes time linear to the size of the js file. On my machine thats about 10ms for 100k of js.
With Opal that is no longer true, the work basically starts after that inital evaluation.
For a 730k file of opal code, one has to expect the (on my machine) 70ms for evaluating the "literal" javascript and now add the booting of the ruby runtime along with time for creating all those classes, methods etc and things added by metaprogramming.
That in the example of the 730k File another ~360ms in addition to the initial 70ms for the "literal" js.
That makes the boot of opal 6 times slower vs "literal" js code.
In general it seem valuable to defer the creation of classes/methods until they are needed.
The time for a class creation, depending on complexity of the class, is expected to take only a few ms, maybe 1 to 5ms.
Creating the class "on demand" introduces a only small delay. But creating all classes at boot time creates a big delay increasing TTI a lot.
<img width="121" alt="Bildschirmfoto 2019-10-15 um 17 26 56" src="https://user-images.githubusercontent.com/30776880/68039732-08f0ba00-fccd-11e9-872f-3b91689e1b9f.png">
Measured times and sizes for the above mentioned example.
Green column literal js, red column opal code.
Top row time in ms, bottom row code size uncompressed but minified in kb.

**Comment by janbiedermann** [comment]
In es6_modules_1_1 `autoload` now also works as dynamic method call and outside of classes/modules. Probably it can be used for corelib autoloading too, i did not yet try.

At the same time `require_tree 'bla', :autoload` now does nothing in the browser, instead a autoloader would be responsible.
At the same time now opal-zeitwerk is available as autoloader.

Performance:
- requiring opal-zeitwerk introduces a ~5ms penalty,
- setting up autoloading roughly too, depending on the amount of autoloaded Opal.modules paths and their contents.
- autoloading constants with opal-zeitwerk its pretty fast, as it uses the standard Opal mechanisms (improved in es6_modules_1_1 to be compatible with standard ruby/zeitwerk style)

Introducing opal-zeitwerk is a big win for large opal projects.

**Comment by hmdne** [comment]
To update - Opal 1.3.0 autoloads a few corelib classes
